### PR TITLE
Tsk 1938

### PR DESCRIPTION
**Pull request by ryzheboka**
https://sonarcloud.io/summary/new_code?id=ryzheboka_taskana&branch=TSK-1938

---
Release Notes:

```
TSK-1937: added AfterRequestChangesProvider which allows customized behavior after changes have been requested on a given Task
```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [x] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: lib/taskana-core/src/main/java/pro/taskana/spi/task/api/AfterRequestChangesProvider.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+   * active {@linkplain TaskanaEngine} for later usage.
+   *
+   * @param taskanaEngine the active {@linkplain TaskanaEngine} which is initialized for this
+   *     installation
+   */
+  void initialize(TaskanaEngine taskanaEngine);
+
+  /**
+   * Perform any action after have been requested on a {@linkplain Task} through {@linkplain
+   * pro.taskana.task.api.TaskService#requestChanges(String)}.
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/spi/task/api/AfterRequestChangesProvider.java:26]: Something is missing in this description.

Perform any action after [a review] ha[s] been requested

File: lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    @TaskanaInject TaskService taskService;
+
+    @WithAccessId(user = "user-1-1")
+    @Test
+    void should_ReturnModifiedTask_When_SpiModifiesAndTransfersTask() throws Exception {
+      Task task = createTaskInReviewByUser("user-1-1").buildAndStore(taskService);
+
+      Task result = taskService.requestChanges(task.getId());
+
+      assertThat(result.getCustomField(TaskCustomField.CUSTOM_3)).isEqualTo("bla");
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:160]: Should we make "bla" a constant?

We didn't do this in the other test aswell.

File: lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+  @WithServiceProvider(
+      serviceProviderInterface = AfterRequestChangesProvider.class,
+      serviceProviders = ClassificationUpdater.class)
+  class SpiModifiesTaskAndClassification {
+
+    @TaskanaInject TaskService taskService;
+
+    @WithAccessId(user = "user-1-1", groups = "businessadmin")
+    @Test
+    void should_ReturnUpdateMultipleEntities_When_SpiModifiesTask() throws Exception {
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:188]: this name makes no sense to me :sweat_smile: .
Same applies to the other test class.

Comment (mustaphazorgati) [reply lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:188]: Suggestion: `should_PersistDifferentEntityChanges_When_SpiModifiesMoreThanTheTask` or something like that?

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

File: lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java (hunk 2 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:160]: I will create the constant in both tests

File: lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java (hunk 3 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:188]: What do you think of `should_ChangeMultipleEntities_When_SpiModifiesMoreThanTheTask`? It sounds a little bit simpler to me.

File: lib/taskana-core/src/main/java/pro/taskana/spi/task/api/AfterRequestChangesProvider.java (hunk 1 above)
Comment (ryzheboka) [reply lib/taskana-core/src/main/java/pro/taskana/spi/task/api/AfterRequestChangesProvider.java:26]: I will complete the description :)

File: lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java (hunk 3 above)
Comment (mustaphazorgati) [reply lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:188]: sure

File: lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+import pro.taskana.testapi.security.WithAccessId;
+import pro.taskana.workbasket.api.WorkbasketPermission;
+import pro.taskana.workbasket.api.WorkbasketService;
+import pro.taskana.workbasket.api.models.WorkbasketSummary;
+
+@TaskanaIntegrationTest
+public class RequestChangesWithSpiAccTest {
+
+  private static final String NEW_WORKBASKET_KEY = "W1000";
+  private static final String NEW_CUSTOM_3_VALUE = "bla";
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/requestchanges/RequestChangesWithSpiAccTest.java:44]: This is nitpicking: Technically this constant should be located within the SPI.

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

**Review by mustaphazorgati — APPROVED** [review]

**Pull request merged**